doc/how-to-analyze-haskell-code-performance.rst

```diff
@@ -0,0 +1,146 @@
+      `"eventlog" report <https://downloads.haskell.org/ghc/latest/docs/users_guide/runtime_control.html#rts-eventlog>`__
+      ``<app-name>.eventlog``, which contains a lot more details and can show you resource usage over time, and can
+      be converted to JSON with `hs-speedscope <https://github.com/mpickering/hs-speedscope>`__
+      to be visualized with `Speedscope <https://speedscope.app>`__.
+
+Finally, visualize this JSON report ``my-app.prof`` and analyze it for performance bottlenecks.
+One popular open-source
+`flame graph <https://www.brendangregg.com/flamegraphs.html>`__
+visualizer is
+`Speedscope <https://speedscope.app>`__,
+which runs in the browser and can open this JSON file directly.
+See the
+`Haskell Optimization Handbook <https://haskell.foundation/hs-opt-handbook.github.io>`__
+on how to optimize your code based on the profiling results afterwards.
+
+So far, we've only used a single GHC compiler option to enable profiling
+in general for your application.
+However, you can pass a lot more compiler flags to GHC via the ``cabal.project.local`` file:
+
+.. code-block:: cabal
+
+    profiling: True
+    profiling-detail: late-toplevel
+    program-options
+      ghc-options:
+        -fprof-auto
+        -fno-prof-count-entries
+        -fprof-auto-calls
+
+Add these recommended profiling options to control where and when
+performance measuring code is inserted into your application
+The setting ``profiling-detail: late-toplevel`` instructs GHC to use so-called
+`late-cost-center profiling <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#ghc-flag--fprof-late>`__
+and insert measuring code only after important optimisations have been applied to your application code.
+This reduces the performance slow-down of profiling itself and gives you more realistic measurements.
+
+The ``ghc-options`` setting allows you to further control which functions and other bindings
+the GHC compiler should profile.
+You can find more information and further options in the
+`GHC "cost-center" guide <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#automatically-placing-cost-centres>`__.
+and the
+`GHC profiling compiler options <https://downloads.haskell.org/ghc/latest/docs/users_guide/profiling.html#compiler-options-for-profiling>`__
+section.
+
+Profiling your dependencies too
+-------------------------------
+
+The setup so far only profiles your main application, which is usually what you want.
+This happens by default, because Cabal command line options only apply to local packages
+and dependencies are usually not local.
+However, the bottlenecks may be in your dependencies, so you would want to profile those too.
+
+First, to enable ``late``-cost-center profiling of all packages/dependencies in your project,
+add the following to your project’s ``cabal.project.local`` file:
+
+.. code-block:: cabal
+
+    package *
+        profiling: true
+        profiling-detail: late
```

Comment:
As you mentioned `program-options` above, you can mention here that further options might be passed via:
```
package *
  profiling: true
  profiling-detail: late
  ghc-options:
    -fno-prof-count-entries
```

Reply:
Also perhaps I would mention that one should add:
```
package <my-dependency>
  profiling: true
  profiling-detail: late
  ghc-options:
    -fno-prof-count-entries
```

to the cabal.project.local, and *if one wants it to apply to all dependencies* then use `package *`. See the program-options section https://cabal.readthedocs.io/en/latest/cabal-project-description-file.html#cfg-section-program-options-program-options

- `program-options`: all local packages
- `package blah`: only package `blah` (be it local or remote)
- `package *`: all local + remote packages

Reply:
This is now a bit wrong. `package <package-name` applies to a package, be it local or remote.

Reply:
Fixed